File: combinationSum.py
def combinationSum(candidates,target): # TC O(N^(T/M+1)) // SC O(T/M)
	results = []

	def backtrack(remain, comb, start):
	    if remain == 0:                                      # goal
	        # make a deep copy of the current combination
	        results.append(list(comb))
	        return
	    elif remain < 0:
	        # exceed the scope, stop exploration.
	        return
	    for i in range(start, len(candidates)):
	        # add the number into the combination
	        comb.append(candidates[i])                          
	        # give the current number another chance, rather than moving on
	        backtrack(remain - candidates[i], comb, i)                  # choose
	        # backtrack, remove the number from the combination
	        comb.pop()
	backtrack(target, [], 0)
	return results

print(combinationSum([2,3,6,7],7))

File: combinationSum.java
import java.util.*;
class Solution {
    public void backtrack(int[] candidates,List<List<Integer>> answer, int target, List<Integer> combination, int start){
        if (target==0){
            answer.add(new ArrayList<Integer>(combination));
            return;
        } else if (target<0){
            return;
        }
        for (int i=start; i<candidates.length; i++){
            combination.add(candidates[i]);
            backtrack(candidates,answer,target-candidates[i],combination,i);
            combination.remove(combination.size()-1);
        }
        
    }
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<List<Integer>> answer=new ArrayList<>();
        List<Integer> combination=new ArrayList<>();
        backtrack(candidates,answer,target,combination,0);
        return answer;
    }
}

class combinationSum{
    public static void main(String[] args) {
        int[] candidates={2,3,6,7};
        int target=7;
        Solution s= new Solution();
        System.out.println(s.combinationSum(candidates,target));
    }
}